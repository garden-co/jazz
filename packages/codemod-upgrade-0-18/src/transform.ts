import {
  Project,
  SyntaxKind,
  Node,
  PropertyAccessExpression,
  CallExpression,
  BinaryExpression,
  ElementAccessExpression,
  DeleteExpression,
  SourceFile,
  VariableStatement,
  VariableDeclaration,
  ObjectBindingPattern,
} from "ts-morph";
import fs from "node:fs";

// Helper function to check if a property access has optional chaining
function hasOptionalChaining(node: PropertyAccessExpression): boolean {
  return node.hasQuestionDotToken();
}

// Helper function to check if an element access has optional chaining
function hasOptionalChainingElement(node: ElementAccessExpression): boolean {
  return node.hasQuestionDotToken();
}

// Helper function to check if any node in a chain has optional chaining
function hasOptionalChainingInChain(node: Node): boolean {
  if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
    const propAccess = node as PropertyAccessExpression;
    return (
      hasOptionalChaining(propAccess) ||
      hasOptionalChainingInChain(propAccess.getExpression())
    );
  } else if (node.getKind() === SyntaxKind.ElementAccessExpression) {
    const elementAccess = node as ElementAccessExpression;
    return (
      hasOptionalChainingElement(elementAccess) ||
      hasOptionalChainingInChain(elementAccess.getExpression())
    );
  }
  return false;
}

// Helper function to get the text with optional chaining preserved
function getTextWithOptionalChaining(node: Node): string {
  if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
    const propAccess = node as PropertyAccessExpression;
    const expressionText = getTextWithOptionalChaining(
      propAccess.getExpression(),
    );
    if (hasOptionalChaining(propAccess)) {
      return `${expressionText}?.${propAccess.getNameNode().getText()}`;
    }
    return `${expressionText}.${propAccess.getNameNode().getText()}`;
  } else if (node.getKind() === SyntaxKind.ElementAccessExpression) {
    const elementAccess = node as ElementAccessExpression;
    const expressionText = getTextWithOptionalChaining(
      elementAccess.getExpression(),
    );
    const argument = elementAccess.getArgumentExpression();
    if (argument) {
      return `${expressionText}[${argument.getText()}]`;
    }
    return expressionText;
  }
  return node.getText();
}

// Helper function to create property access text with optional chaining preserved
function createPropertyAccessText(
  object: Node,
  propertyName: string,
  useOptionalChaining: boolean = false,
): string {
  const objectText = getTextWithOptionalChaining(object);
  if (useOptionalChaining) {
    // If the property name contains dots, only apply optional chaining to the first part
    if (propertyName.includes(".")) {
      const [firstPart, ...rest] = propertyName.split(".");
      return `${objectText}?.${firstPart}.${rest.join(".")}`;
    }
    return `${objectText}?.${propertyName}`;
  }
  return `${objectText}.${propertyName}`;
}

export function transformFile(sourceFile: SourceFile): string {
  // 1. Transform property access patterns
  transformPropertyAccess(sourceFile);

  // 2. Transform destructuring assignments
  transformDestructuringAssignments(sourceFile);

  // 3. Transform _refs.property to $jazz.refs.property
  transformRefsAccess(sourceFile);

  // 4. Transform _edits.property to $jazz.getEdits().property
  transformEditsAccess(sourceFile);

  // 5. Transform method calls
  transformMethodCalls(sourceFile);

  // 6. Transform array operations
  transformArrayOperations(sourceFile);

  // 7. Transform property assignments
  transformPropertyAssignments(sourceFile);

  // 8. Transform array index assignments
  transformArrayIndexAssignments(sourceFile);

  // 9. Transform delete statements
  transformDeleteStatements(sourceFile);

  // 10. Transform splice operations
  transformSpliceOperations(sourceFile);

  // 11. Remove .castAs() calls
  removeCastAsCalls(sourceFile);

  return sourceFile.getFullText();
}

function isJazzValue(object: Node) {
  // When transformed to text the import path is shown
  // Then we check that this is something that comes from jazz-tools
  return (
    object.getType().getText().includes("jazz-tools") ||
    object.getType().isAny()
  );
}

function transformPropertyAccess(sourceFile: SourceFile) {
  // Transform obj.id to obj.$jazz.id
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const propertyName = propertyAccess.getNameNode();

      if (node.getType().isString() || node.getType().isNumber()) {
        return;
      }

      if (propertyName.getText() === "id") {
        const object = propertyAccess.getExpression();

        if (
          !isJazzValue(object) ||
          object.getType().getText().includes(".Ref<")
        ) {
          return;
        }

        if (object.getText().includes(".$jazz")) {
          return;
        }

        if (object.getKind() === SyntaxKind.PropertyAccessExpression) {
          const objProperty = (
            object as PropertyAccessExpression
          ).getNameNode();
          if (objProperty.getText() === "$jazz") {
            return; // Already transformed
          }
        }

        // Preserve optional chaining
        const hasOptional = hasOptionalChaining(propertyAccess);
        const newText = createPropertyAccessText(
          object,
          "$jazz.id",
          hasOptional,
        );
        propertyAccess.replaceWithText(newText);
      }
    }
  });

  // Transform obj._owner to obj.$jazz.owner
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const propertyName = propertyAccess.getNameNode();

      if (propertyName.getText() === "_owner") {
        const object = propertyAccess.getExpression();
        const hasOptional = hasOptionalChaining(propertyAccess);
        let newText = createPropertyAccessText(
          object,
          "$jazz.owner",
          hasOptional,
        );

        propertyAccess.replaceWithText(newText);
      }
    }
  });

  // Transform obj._type to obj.$type$
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const propertyName = propertyAccess.getNameNode();

      if (propertyName.getText() === "_type") {
        const object = propertyAccess.getExpression();
        const hasOptional = hasOptionalChaining(propertyAccess);
        const newText = createPropertyAccessText(object, "$type$", hasOptional);
        propertyAccess.replaceWithText(newText);
      }
    }
  });

  // Transform obj._createdAt to obj.$jazz.createdAt
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const propertyName = propertyAccess.getNameNode();

      if (propertyName.getText() === "_createdAt") {
        const object = propertyAccess.getExpression();
        const hasOptional = hasOptionalChaining(propertyAccess);
        const newText = createPropertyAccessText(
          object,
          "$jazz.createdAt",
          hasOptional,
        );
        propertyAccess.replaceWithText(newText);
      } else if (propertyName.getText() === "_lastUpdatedAt") {
        const object = propertyAccess.getExpression();
        const hasOptional = hasOptionalChaining(propertyAccess);
        const newText = createPropertyAccessText(
          object,
          "$jazz.lastUpdatedAt",
          hasOptional,
        );
        propertyAccess.replaceWithText(newText);
      }
    }
  });
}

function transformDestructuringAssignments(sourceFile: SourceFile) {
  // Transform destructuring assignments like const { id } = coValue to const { id } = coValue.$jazz
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.VariableStatement) {
      const variableStatement = node as VariableStatement;
      const declarationList = variableStatement.getDeclarationList();

      declarationList.getDeclarations().forEach((declaration) => {
        const nameNode = declaration.getNameNode();
        const initializer = declaration.getInitializer();

        // Check if this is an object destructuring pattern
        if (
          nameNode.getKind() === SyntaxKind.ObjectBindingPattern &&
          initializer
        ) {
          const objectPattern = nameNode as ObjectBindingPattern;

          // Check if any of the destructured properties are Jazz properties we need to transform
          const hasJazzProperties = objectPattern
            .getElements()
            .some((element) => {
              const propertyName = element.getPropertyNameNode();
              if (propertyName) {
                const name = propertyName.getText();
                return (
                  name === "id" ||
                  name === "_owner" ||
                  name === "_type" ||
                  name === "_createdAt" ||
                  name === "_lastUpdatedAt"
                );
              }
              return false;
            });

          if (hasJazzProperties && isJazzValue(initializer)) {
            // Check if initializer already includes $jazz
            if (initializer.getText().includes(".$jazz")) {
              return;
            }

            // Transform based on the properties being destructured
            const elements = objectPattern.getElements();
            const needsJazzTransform = elements.some((element) => {
              const propertyName = element.getPropertyNameNode();
              if (propertyName) {
                const name = propertyName.getText();
                return (
                  name === "id" ||
                  name === "_owner" ||
                  name === "_createdAt" ||
                  name === "_lastUpdatedAt"
                );
              }
              return false;
            });

            const needsTypeTransform = elements.some((element) => {
              const propertyName = element.getPropertyNameNode();
              if (propertyName) {
                const name = propertyName.getText();
                return name === "_type";
              }
              return false;
            });

            if (needsJazzTransform && needsTypeTransform) {
              // Mixed case: split into multiple statements
              const jazzElements = elements.filter((element) => {
                const propertyName = element.getPropertyNameNode();
                if (propertyName) {
                  const name = propertyName.getText();
                  return (
                    name === "id" ||
                    name === "_owner" ||
                    name === "_createdAt" ||
                    name === "_lastUpdatedAt"
                  );
                }
                return false;
              });

              const typeElements = elements.filter((element) => {
                const propertyName = element.getPropertyNameNode();
                if (propertyName) {
                  const name = propertyName.getText();
                  return name === "_type";
                }
                return false;
              });

              const otherElements = elements.filter((element) => {
                const propertyName = element.getPropertyNameNode();
                if (propertyName) {
                  const name = propertyName.getText();
                  return ![
                    "id",
                    "_owner",
                    "_type",
                    "_createdAt",
                    "_lastUpdatedAt",
                  ].includes(name);
                }
                return true;
              });

              // Build new variable statements
              let newStatements: string[] = [];

              if (jazzElements.length > 0) {
                const jazzElementsText = jazzElements
                  .map((el) => {
                    const propertyName = el.getPropertyNameNode();
                    const name = el.getName();
                    if (propertyName && name) {
                      const propText = propertyName.getText();
                      if (propText === "_owner") {
                        return name !== "owner" ? `owner: ${name}` : "owner";
                      } else if (propText === "_createdAt") {
                        return name !== "createdAt"
                          ? `createdAt: ${name}`
                          : "createdAt";
                      } else if (propText === "_lastUpdatedAt") {
                        return name !== "lastUpdatedAt"
                          ? `lastUpdatedAt: ${name}`
                          : "lastUpdatedAt";
                      }
                      return el.getText();
                    }
                    return el.getText();
                  })
                  .join(", ");
                newStatements.push(
                  `const { ${jazzElementsText} } = ${initializer.getText()}.$jazz`,
                );
              }

              if (typeElements.length > 0) {
                const typeElementsText = typeElements
                  .map((el) => {
                    const name = el.getName();
                    return name !== "$type$" ? `$type$: ${name}` : "$type$";
                  })
                  .join(", ");
                newStatements.push(
                  `const { ${typeElementsText} } = ${initializer.getText()}`,
                );
              }

              if (otherElements.length > 0) {
                const otherElementsText = otherElements
                  .map((el) => el.getText())
                  .join(", ");
                newStatements.push(
                  `const { ${otherElementsText} } = ${initializer.getText()}`,
                );
              }

              variableStatement.replaceWithText(
                newStatements.join(";\n") + ";",
              );
            } else if (needsJazzTransform) {
              // Transform property names for $jazz access
              const transformedElements = elements
                .map((element) => {
                  const propertyName = element.getPropertyNameNode();
                  const name = element.getName();
                  if (propertyName && name) {
                    const propText = propertyName.getText();
                    if (propText === "_owner") {
                      return name !== "owner" ? `owner: ${name}` : "owner";
                    } else if (propText === "_createdAt") {
                      return name !== "createdAt"
                        ? `createdAt: ${name}`
                        : "createdAt";
                    } else if (propText === "_lastUpdatedAt") {
                      return name !== "lastUpdatedAt"
                        ? `lastUpdatedAt: ${name}`
                        : "lastUpdatedAt";
                    }
                    return element.getText();
                  }
                  return element.getText();
                })
                .join(", ");

              const newText = `const { ${transformedElements} } = ${initializer.getText()}.$jazz`;
              variableStatement.replaceWithText(newText);
            } else if (needsTypeTransform) {
              // Transform _type to $type$
              const transformedElements = elements
                .map((element) => {
                  const propertyName = element.getPropertyNameNode();
                  const name = element.getName();
                  if (propertyName && name) {
                    const propText = propertyName.getText();
                    if (propText === "_type") {
                      return name !== "$type$" ? `$type$: ${name}` : "$type$";
                    }
                    return element.getText();
                  }
                  return element.getText();
                })
                .join(", ");

              const newText = `const { ${transformedElements} } = ${initializer.getText()}`;
              variableStatement.replaceWithText(newText);
            }
          }
        }
      });
    }
  });
}

function transformRefsAccess(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const object = propertyAccess.getExpression();

      if (object.getKind() === SyntaxKind.PropertyAccessExpression) {
        const objProperty = (object as PropertyAccessExpression).getNameNode();
        if (objProperty.getText() === "_refs") {
          const baseObject = (
            object as PropertyAccessExpression
          ).getExpression();
          const property = propertyAccess.getNameNode();
          const hasOptional = hasOptionalChaining(propertyAccess);
          const newText = createPropertyAccessText(
            baseObject,
            `$jazz.refs.${property.getText()}`,
            hasOptional,
          );
          propertyAccess.replaceWithText(newText);
        }
      }
    }
  });
}

function transformEditsAccess(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = node as PropertyAccessExpression;
      const object = propertyAccess.getExpression();

      if (object.getKind() === SyntaxKind.PropertyAccessExpression) {
        const objProperty = (object as PropertyAccessExpression).getNameNode();
        if (objProperty.getText() === "_edits") {
          const baseObject = (
            object as PropertyAccessExpression
          ).getExpression();
          const property = propertyAccess.getNameNode();
          const hasOptional = hasOptionalChaining(propertyAccess);
          const newText = createPropertyAccessText(
            baseObject,
            `$jazz.getEdits().${property.getText()}`,
            hasOptional,
          );
          propertyAccess.replaceWithText(newText);
        }
      }
    }
  });
}

function transformMethodCalls(sourceFile: SourceFile) {
  const methodNames = [
    "waitForSync",
    "waitForAllCoValuesSync",
    "subscribe",
    "ensureLoaded",
    "applyDiff",
  ];

  methodNames.forEach((methodName) => {
    sourceFile.forEachDescendant((node) => {
      if (node.getKind() === SyntaxKind.CallExpression) {
        const callExpression = node as CallExpression;
        const callee = callExpression.getExpression();

        if (callee.getKind() === SyntaxKind.PropertyAccessExpression) {
          const property = (callee as PropertyAccessExpression).getNameNode();

          if (property.getText() === methodName) {
            const baseObject = (
              callee as PropertyAccessExpression
            ).getExpression();

            if (baseObject.getText().includes(".$jazz")) {
              return;
            }

            const args = callExpression
              .getArguments()
              .map((arg) => arg.getText())
              .join(", ");

            // Preserve optional chaining from the method call
            const hasOptional = hasOptionalChaining(
              callee as PropertyAccessExpression,
            );
            const newText = createPropertyAccessText(
              baseObject,
              `$jazz.${methodName}(${args})`,
              hasOptional,
            );
            callExpression.replaceWithText(newText);
          }
        }
      }
    });
  });
}

function transformArrayOperations(sourceFile: SourceFile) {
  const arrayOperationNames = ["push", "unshift", "shift", "pop"];

  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.CallExpression) {
      const callExpression = node as CallExpression;
      const callee = callExpression.getExpression();

      if (callee.getKind() === SyntaxKind.PropertyAccessExpression) {
        const property = (callee as PropertyAccessExpression).getNameNode();
        const propertyName = property.getText();

        if (arrayOperationNames.includes(propertyName)) {
          const baseObject = (
            callee as PropertyAccessExpression
          ).getExpression();

          if (!isJazzValue(baseObject)) {
            return;
          }

          if (baseObject.getText().includes(".$jazz")) {
            return;
          }

          const args = callExpression
            .getArguments()
            .map((arg) => arg.getText())
            .join(", ");

          // Preserve optional chaining from the method call
          const hasOptional = hasOptionalChaining(
            callee as PropertyAccessExpression,
          );
          const newText = createPropertyAccessText(
            baseObject,
            `$jazz.${propertyName}(${args})`,
            hasOptional,
          );
          callExpression.replaceWithText(newText);
        }
      }
    }
  });
}

function transformPropertyAssignments(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.BinaryExpression) {
      const binaryExpression = node as BinaryExpression;
      if (binaryExpression.getOperatorToken().getText() === "=") {
        const left = binaryExpression.getLeft();
        const right = binaryExpression.getRight();

        if (left.getKind() === SyntaxKind.PropertyAccessExpression) {
          const leftPropertyAccess = left as PropertyAccessExpression;

          // Skip if it's already a $jazz.set call
          const leftObject = leftPropertyAccess.getExpression();

          if (!isJazzValue(leftObject)) {
            return;
          }

          if (leftObject.getKind() === SyntaxKind.PropertyAccessExpression) {
            const objProperty = (
              leftObject as PropertyAccessExpression
            ).getNameNode();
            if (objProperty.getText() === "$jazz") {
              return;
            }
          }

          const propertyName = leftPropertyAccess.getNameNode().getText();
          const object = leftPropertyAccess.getExpression();

          // Handle undefined assignments as delete
          if (
            right.getKind() === SyntaxKind.Identifier &&
            right.getText() === "undefined"
          ) {
            const hasOptional = hasOptionalChaining(leftPropertyAccess);
            const newText = createPropertyAccessText(
              object,
              `$jazz.delete("${propertyName}")`,
              hasOptional,
            );
            binaryExpression.replaceWithText(newText);
          } else {
            const leftObject = leftPropertyAccess.getExpression();
            const hasOptional = hasOptionalChaining(leftPropertyAccess);
            if (leftObject.getType().isUnion()) {
              const newText = createPropertyAccessText(
                object,
                `$jazz.applyDiff({"${propertyName}": ${right.getText()}})`,
                hasOptional,
              );
              binaryExpression.replaceWithText(newText);
            } else {
              // Handle regular assignments as set
              const newText = createPropertyAccessText(
                object,
                `$jazz.set("${propertyName}", ${right.getText()})`,
                hasOptional,
              );
              binaryExpression.replaceWithText(newText);
            }
          }
        }
      }
    }
  });
}

function transformArrayIndexAssignments(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.BinaryExpression) {
      const binaryExpression = node as BinaryExpression;
      if (binaryExpression.getOperatorToken().getText() === "=") {
        const left = binaryExpression.getLeft();
        const right = binaryExpression.getRight();

        if (left.getKind() === SyntaxKind.ElementAccessExpression) {
          const elementAccess = left as ElementAccessExpression;
          const object = elementAccess.getExpression();
          const index = elementAccess.getArgumentExpression();

          if (!isJazzValue(object)) {
            return;
          }

          if (index) {
            // For array index assignments, we need to check if the parent property access has optional chaining
            // Since this is an element access, we'll check the object for optional chaining
            const hasOptional = hasOptionalChainingInChain(object);
            const newText = createPropertyAccessText(
              object,
              `$jazz.set(${index.getText()}, ${right.getText()})`,
              hasOptional,
            );
            binaryExpression.replaceWithText(newText);
          }
        }
      }
    }
  });
}

function transformDeleteStatements(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.DeleteExpression) {
      const deleteExpression = node as DeleteExpression;
      const argument = deleteExpression.getExpression();

      if (argument.getKind() === SyntaxKind.PropertyAccessExpression) {
        const property = (argument as PropertyAccessExpression).getNameNode();
        const object = (argument as PropertyAccessExpression).getExpression();

        if (!isJazzValue(object)) {
          return;
        }

        const hasOptional = hasOptionalChaining(
          argument as PropertyAccessExpression,
        );
        const newText = createPropertyAccessText(
          object,
          `$jazz.delete("${property.getText()}")`,
          hasOptional,
        );
        deleteExpression.replaceWithText(newText);
      }
    }
  });
}

function transformSpliceOperations(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.CallExpression) {
      const callExpression = node as CallExpression;
      const callee = callExpression.getExpression();

      if (callee.getKind() === SyntaxKind.PropertyAccessExpression) {
        const property = (callee as PropertyAccessExpression).getNameNode();
        if (property.getText() === "splice") {
          const baseObject = (
            callee as PropertyAccessExpression
          ).getExpression();

          if (!isJazzValue(baseObject)) {
            return;
          }

          if (baseObject.getText().includes(".$jazz")) {
            return;
          }

          const args = callExpression
            .getArguments()
            .map((arg) => arg.getText())
            .join(", ");

          // Preserve optional chaining from the method call
          const hasOptional = hasOptionalChaining(
            callee as PropertyAccessExpression,
          );
          const newText = createPropertyAccessText(
            baseObject,
            `$jazz.splice(${args})`,
            hasOptional,
          );
          callExpression.replaceWithText(newText);
        }
      }
    }
  });
}

function removeCastAsCalls(sourceFile: SourceFile) {
  sourceFile.forEachDescendant((node) => {
    if (node.getKind() === SyntaxKind.CallExpression) {
      const callExpression = node as CallExpression;
      const callee = callExpression.getExpression();

      if (!isJazzValue(callee)) {
        return;
      }

      if (callee.getKind() === SyntaxKind.PropertyAccessExpression) {
        const property = (callee as PropertyAccessExpression).getNameNode();
        if (property.getText() === "castAs") {
          const baseObject = (
            callee as PropertyAccessExpression
          ).getExpression();

          // Replace the entire call expression with just the base object
          // This removes .castAs(Group) and keeps just the object
          callExpression.replaceWithText(baseObject.getText());
        }
      }
    }
  });
}

// Main function to run the transform
function singleRun(projectPath: string) {
  let project: Project;

  if (fs.existsSync(`${projectPath}/tsconfig.json`)) {
    project = new Project({
      tsConfigFilePath: `${projectPath}/tsconfig.json`,
    });
  } else {
    project = new Project();

    // Check if projectPath ends with a supported file extension
    const supportedExtensions = [".ts", ".tsx"];
    const hasSupportedExtension = supportedExtensions.some((ext) =>
      projectPath.endsWith(ext),
    );

    if (hasSupportedExtension) {
      // If it's a specific file, add it directly
      project.addSourceFilesAtPaths(projectPath);
    } else {
      // If it's a directory, add with glob pattern
      project.addSourceFilesAtPaths(`${projectPath}/**/*.{ts,tsx}`);
    }
  }

  const sourceFiles = project.getSourceFiles();

  let changed = false;

  sourceFiles.forEach((sourceFile) => {
    if (sourceFile.getFilePath().includes("node_modules")) {
      return;
    }

    try {
      const originalText = sourceFile.getFullText();
      const transformedText = transformFile(sourceFile);

      if (originalText !== transformedText) {
        sourceFile.saveSync();
        console.log(`Transformed: ${sourceFile.getFilePath()}`);
        changed = true;
      }
    } catch (error) {
      console.error(`Error transforming ${sourceFile.getFilePath()}:`, error);
    }
  });

  return changed;
}

// Run multiple times, because the resolved types gets fixed after each run
export function runTransform(projectPath: string): void {
  let runCount = 1;

  while (singleRun(projectPath)) {
    runCount++;

    if (runCount > 5) {
      console.log("Ran 5 times, stopping");
      break;
    }
  }

  console.log(`Ran ${runCount} times`);
}
