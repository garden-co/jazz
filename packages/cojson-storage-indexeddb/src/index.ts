export {
  internal_setDatabaseName,
  getIndexedDBStorage,
} from "./idbNode.js";
