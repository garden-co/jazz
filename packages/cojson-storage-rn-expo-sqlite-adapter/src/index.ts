export { ExpoSQLiteAdapter } from "./expoSQLiteAdapter.js";
