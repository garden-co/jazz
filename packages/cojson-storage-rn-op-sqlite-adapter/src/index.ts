export { OPSQLiteAdapter } from "./opSQLiteAdapter.js";
