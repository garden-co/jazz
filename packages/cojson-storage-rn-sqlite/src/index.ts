export {
  SQLiteReactNative,
  SQLiteReactNative as SQLiteStorage,
} from "./sqlite-react-native.js";
