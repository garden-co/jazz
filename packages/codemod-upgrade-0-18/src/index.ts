#!/usr/bin/env node

import { runTransform } from "./transform.js";

const args = process.argv.slice(2);
const projectPath = args[0] || process.cwd();

console.log(`Running Jazz API migration transform on: ${projectPath}`);
runTransform(projectPath);
console.log("Transform completed!");
