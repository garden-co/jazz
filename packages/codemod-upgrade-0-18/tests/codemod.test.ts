import { expect, onTestFinished, test } from "vitest";
import { runTransform } from "../src/transform";
import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

const currentDir = fileURLToPath(new URL(".", import.meta.url));

test("runTransform", () => {
  const fixturesBefore = readFileSync(
    `${currentDir}/fixtures/test-example.ts`,
    "utf-8",
  );
  runTransform(`${currentDir}/fixtures`);

  const result = readFileSync(
    `${currentDir}/fixtures/test-example.ts`,
    "utf-8",
  );

  onTestFinished(() => {
    writeFileSync(`${currentDir}/fixtures/test-example.ts`, fixturesBefore);
  });

  expect(result).toMatchSnapshot();
});
