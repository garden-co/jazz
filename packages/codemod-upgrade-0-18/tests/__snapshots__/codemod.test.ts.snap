// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runTransform 1`] = `
"// @ts-nocheck

// Test file demonstrating old Jazz API patterns that will be migrated
// This file shows examples of what the codemod will transform

import { Account, co, Group, z } from "jazz-tools";

const Message = co.map({
  text: z.string(),
});
const Chat = co.list(Message);

const Union = co.discriminatedUnion("type", [
  co.map({
    type: z.literal("a"),
    name: z.string(),
  }),
  co.map({
    type: z.literal("b"),
    name: z.string(),
  }),
]);
type Union = co.loaded<typeof Union>;

// Example 1: Property access patterns
function examplePropertyAccess() {
  const chat = Chat.create([], Group.create());
  const chatId = chat.$jazz.id; // Will become: chat.$jazz.id
  const owner = chat.$jazz.owner; // Will become: chat.$jazz.owner
  const raw = chat.$jazz.raw; // Will become: chat.$jazz.raw
  const refs = chat.$jazz.refs.messages; // Will become: chat.$jazz.refs.messages
  const edits = chat.$jazz.getEdits().title; // Will become: chat.$jazz.getEdits().title
  const type = chat.$type$; // Will become: chat.$type$
  const createdAt = chat.$jazz.createdAt; // Will become: chat.$jazz.createdAt
  const lastUpdatedAt = chat.$jazz.lastUpdatedAt; // Will become: chat.$jazz.lastUpdatedAt

  // Should keep conditionals
  const id = chat?.$jazz.id;

  return {
    chatId,
    owner,
    raw,
    refs,
    edits,
    type,
    createdAt,
    lastUpdatedAt,
  };
}

function invalidSimilarCode() {
  const arr = [1, 2, 3];
  arr.push(4);
  arr.splice(0, 1);
  arr[0] = 5;
  delete arr[0];

  const obj: Record<string, string> = {
    name: "John",
    age: "30",
    city: "New York",
    castAs() {
      return this;
    },
  };
  obj.name = "Jane";
  delete obj.age;
  obj.castAs(Group);
}

function invalidSimilarCode2(args: { id: string; message: Message }) {
  return args.id;
}

// Example 2: Property assignments
function examplePropertyAssignments(union: Union) {
  const me = Account.getMe();

  if (me.profile) {
    me.profile.$jazz.set("name", "New Name"); // Will become: me.profile.$jazz.set("name", "New Name")
    me.profile.$jazz.delete("avatar"); // Will become: me.profile.$jazz.delete("avatar")
  }

  const chat = Message.create({ text: "Hello" }, Group.create());
  chat.$jazz.set("text", "My Chat");

  union.$jazz.applyDiff({"name": "John"}); // Will become: union.$jazz.applyDiff({"name": "John"}) because unions do not support set
}

// Example 3: Array operations
function exampleArrayOperations() {
  const chat = Chat.create([], Group.create());
  const message = Message.create({ text: "Hello" }, chat.$jazz.owner);

  chat.$jazz.push(message); // Will become: chat.messages.$jazz.push(message)
  chat.$jazz.splice(0, 1); // Will become: chat.messages.$jazz.splice(0, 1)
  chat.$jazz.unshift(message); // Will become: chat.messages.$jazz.unshift(message)
  chat.$jazz.shift(); // Will become: chat.messages.$jazz.shift()
  chat.$jazz.pop(); // Will become: chat.messages.$jazz.pop()
}

// Example 4: Method calls
async function exampleMethodCalls() {
  const chat = Chat.create([], Group.create());

  await chat.$jazz.waitForSync(); // Will become: await chat.$jazz.waitForSync()
  await chat.$jazz.waitForAllCoValuesSync(); // Will become: await chat.$jazz.waitForAllCoValuesSync()

  const unsubscribe = chat.$jazz.subscribe({}, (updatedChat) => {
      console.log("Chat updated:", updatedChat.$jazz.id);
    }); // Will become: chat.$jazz.subscribe({}, callback)

  const loadedChat = await chat.$jazz.ensureLoaded({
      resolve: { messages: true },
    }); // Will become: chat.$jazz.ensureLoaded(options)
}

// Example 5: Rich text operations
function exampleRichTextOperations() {
  const richText = co.richText().create("Initial text", Group.create());

  richText.$jazz.applyDiff("Updated text"); // Will become: richText.$jazz.applyDiff("Updated text")
}

// Example 6: Complex property chains
function exampleComplexChains() {
  const chat = Chat.create([], Group.create());
  const message = chat[0];

  if (message && message?.$jazz.getEdits().text?.by?.profile?.name) {
    const authorName = message.$jazz.getEdits().text.by.profile.name;
    // Will become: message.$jazz.getEdits()?.text?.by?.profile?.name
  }

  const messageOwner = message?.$jazz.owner; // Will become: message?.$jazz.owner
  const messageRefs = message?.$jazz.refs.text; // Will become: message?.$jazz.refs?.text
}

// Example 7: Conditional expressions
function exampleConditionals() {
  const chat = Chat.create([], Group.create());

  if (chat?.$jazz.getEdits().title) {
    // Will become: chat.$jazz.getEdits()?.title
    console.log("Chat has title edits");
  }

  const hasOwner = chat.$jazz.owner ? true : false; // Will become: chat.$jazz.owner ? true : false
}

// Example 8: Delete operations
function exampleDeleteOperations() {
  const me = Account.getMe();

  me.profile.$jazz.delete("avatar"); // Will become: me.profile.$jazz.delete("avatar")
  me.profile.$jazz.delete("bio"); // Will become: me.profile.$jazz.delete("bio")
}

// Example 9: Type checking
function exampleTypeChecking() {
  const obj = Chat.create([], Group.create());

  if (obj.$type$ === "Group") {
    // Will become: obj.$type$ === 'Group'
    console.log("This is a Group");
  }

  if (obj.$type$ === "Chat") {
    // Will become: obj.$type$ === 'Chat'
    console.log("This is a Chat");
  }
}

// Example 10: Timestamp access
function exampleTimestamps() {
  const message = Message.create({ text: "Hello" }, Group.create());

  const created = message.$jazz.createdAt; // Will become: message.$jazz.createdAt
  const lastEdit = message?.$jazz.getEdits().text?.madeAt; // Will become: message.$jazz.getEdits()?.text?.madeAt
}

// Example 11: Destructuring assignments
function exampleDestructuring() {
  const chat = Chat.create([], Group.create());
  const message = Message.create({ text: "Hello" }, Group.create());

  // Simple id destructuring
  const { id } = chat; // Will become: const { id } = chat.$jazz

  // Multiple properties from $jazz
  const { id: chatId, _owner, _raw, _createdAt } = chat.$jazz // Will become: const { id: chatId, owner: _owner, raw: _raw, createdAt: _createdAt } = chat.$jazz

  // Type property destructuring
  const { _type } = message; // Will become: const { $type$: _type } = message

  // Mixed properties (should split into multiple statements)
  const { id: messageId, owner: messageOwner, raw: messageRaw } = message.$jazz;
    const { $type$: messageType } = message; // Will split into separate statements

  // Regular properties should not be transformed
  const { text } = message; // Should remain unchanged

  return {
    id,
    chatId,
    _owner,
    _raw,
    _createdAt,
    _type,
    messageId,
    messageType,
    messageOwner,
    messageRaw,
    text,
  };
}

export {
  examplePropertyAccess,
  examplePropertyAssignments,
  exampleArrayOperations,
  exampleMethodCalls,
  exampleRichTextOperations,
  exampleComplexChains,
  exampleConditionals,
  exampleDeleteOperations,
  exampleTypeChecking,
  exampleTimestamps,
  exampleDestructuring,
  invalidSimilarCode,
};
"
`;
